File: snippets/design-patterns/factory/FactoryPattern.java
/**
 * Factory Pattern Implementation in Java
 * 
 * This demonstrates three variations of the Factory Pattern:
 * 1. Simple Factory
 * 2. Factory Method
 * 3. Abstract Factory
 */

// ============= SIMPLE FACTORY EXAMPLE =============

// Product interface
interface Product {
    void operation();
}

// Concrete products
class ConcreteProductA implements Product {
    @Override
    public void operation() {
        System.out.println("Operation of ConcreteProductA");
    }
}

class ConcreteProductB implements Product {
    @Override
    public void operation() {
        System.out.println("Operation of ConcreteProductB");
    }
}

// Simple factory
class SimpleFactory {
    public Product createProduct(String type) {
        if (type.equals("A")) {
            return new ConcreteProductA();
        } else if (type.equals("B")) {
            return new ConcreteProductB();
        }
        throw new IllegalArgumentException("Invalid product type: " + type);
    }
}

// Client for Simple Factory
class SimpleFactoryClient {
    public void run() {
        System.out.println("\n=== Simple Factory Example ===");
        SimpleFactory factory = new SimpleFactory();
        
        Product productA = factory.createProduct("A");
        productA.operation();
        
        Product productB = factory.createProduct("B");
        productB.operation();
    }
}

// ============= FACTORY METHOD EXAMPLE =============

// Product interface for Factory Method
interface Transport {
    void deliver();
}

// Concrete products
class Truck implements Transport {
    @Override
    public void deliver() {
        System.out.println("Delivering by land in a truck");
    }
}

class Ship implements Transport {
    @Override
    public void deliver() {
        System.out.println("Delivering by sea in a ship");
    }
}

// Creator abstract class with factory method
abstract class Logistics {
    // Factory method
    public abstract Transport createTransport();
    
    // Business logic that uses the factory method
    public void planDelivery() {
        System.out.println("Preparing for delivery...");
        Transport transport = createTransport();
        System.out.println("Planning route...");
        transport.deliver();
        System.out.println("Delivery complete!");
    }
}

// Concrete creators override factory method
class RoadLogistics extends Logistics {
    @Override
    public Transport createTransport() {
        return new Truck();
    }
}

class SeaLogistics extends Logistics {
    @Override
    public Transport createTransport() {
        return new Ship();
    }
}

// Client for Factory Method
class FactoryMethodClient {
    public void run() {
        System.out.println("\n=== Factory Method Example ===");
        
        System.out.println("\nRoad delivery:");
        Logistics roadLogistics = new RoadLogistics();
        roadLogistics.planDelivery();
        
        System.out.println("\nSea delivery:");
        Logistics seaLogistics = new SeaLogistics();
        seaLogistics.planDelivery();
    }
}

// ============= ABSTRACT FACTORY EXAMPLE =============

// Abstract products
interface Button {
    void render();
    void click();
}

interface Checkbox {
    void render();
    void toggle();
}

// Concrete products for Windows
class WindowsButton implements Button {
    @Override
    public void render() {
        System.out.println("Rendering a Windows button");
    }
    
    @Override
    public void click() {
        System.out.println("Windows button clicked");
    }
}

class WindowsCheckbox implements Checkbox {
    @Override
    public void render() {
        System.out.println("Rendering a Windows checkbox");
    }
    
    @Override
    public void toggle() {
        System.out.println("Windows checkbox toggled");
    }
}

// Concrete products for macOS
class MacOSButton implements Button {
    @Override
    public void render() {
        System.out.println("Rendering a macOS button");
    }
    
    @Override
    public void click() {
        System.out.println("macOS button clicked");
    }
}

class MacOSCheckbox implements Checkbox {
    @Override
    public void render() {
        System.out.println("Rendering a macOS checkbox");
    }
    
    @Override
    public void toggle() {
        System.out.println("macOS checkbox toggled");
    }
}

// Abstract factory interface
interface GUIFactory {
    Button createButton();
    Checkbox createCheckbox();
}

// Concrete factories
class WindowsFactory implements GUIFactory {
    @Override
    public Button createButton() {
        return new WindowsButton();
    }
    
    @Override
    public Checkbox createCheckbox() {
        return new WindowsCheckbox();
    }
}

class MacOSFactory implements GUIFactory {
    @Override
    public Button createButton() {
        return new MacOSButton();
    }
    
    @Override
    public Checkbox createCheckbox() {
        return new MacOSCheckbox();
    }
}

// Client code that works with factories and products
class Application {
    private Button button;
    private Checkbox checkbox;
    
    public Application(GUIFactory factory) {
        button = factory.createButton();
        checkbox = factory.createCheckbox();
    }
    
    public void createUI() {
        System.out.println("Creating UI components...");
        button.render();
        checkbox.render();
    }
    
    public void clickButton() {
        button.click();
    }
    
    public void toggleCheckbox() {
        checkbox.toggle();
    }
}

// Client for Abstract Factory
class AbstractFactoryClient {
    public void run() {
        System.out.println("\n=== Abstract Factory Example ===");
        
        // Simulate different OS environments
        createApplicationFor("Windows");
        createApplicationFor("macOS");
    }
    
    private void createApplicationFor(String os) {
        System.out.println("\nCreating app for " + os);
        
        GUIFactory factory;
        if (os.equals("Windows")) {
            factory = new WindowsFactory();
        } else if (os.equals("macOS")) {
            factory = new MacOSFactory();
        } else {
            throw new IllegalArgumentException("Unknown OS: " + os);
        }
        
        Application app = new Application(factory);
        app.createUI();
        app.clickButton();
        app.toggleCheckbox();
    }
}

// Main class to run all examples
public class FactoryPattern {
    public static void main(String[] args) {
        System.out.println("Factory Pattern Demonstrations");
        System.out.println("==============================");
        
        // Run Simple Factory example
        new SimpleFactoryClient().run();
        
        // Run Factory Method example
        new FactoryMethodClient().run();
        
        // Run Abstract Factory example
        new AbstractFactoryClient().run();
    }
} 

File: snippets/design-patterns/factory/factory_pattern.py
#!/usr/bin/env python3
"""
Factory Pattern Implementation in Python

This demonstrates three variations of the Factory Pattern:
1. Simple Factory
2. Factory Method
3. Abstract Factory
"""

from abc import ABC, abstractmethod
import sys

# ============= SIMPLE FACTORY EXAMPLE =============

class Product(ABC):
    """Product interface"""
    @abstractmethod
    def operation(self):
        pass

class ConcreteProductA(Product):
    """Concrete product A"""
    def operation(self):
        return "Result of ConcreteProductA operation"

class ConcreteProductB(Product):
    """Concrete product B"""
    def operation(self):
        return "Result of ConcreteProductB operation"

class SimpleFactory:
    """Simple Factory implementation"""
    @staticmethod
    def create_product(product_type):
        """Creates a product based on type"""
        if product_type == "A":
            return ConcreteProductA()
        elif product_type == "B":
            return ConcreteProductB()
        else:
            raise ValueError(f"Product type {product_type} not recognized")

# Client code for Simple Factory
def simple_factory_client():
    """Demo of Simple Factory Pattern"""
    print("\n=== Simple Factory Example ===")
    factory = SimpleFactory()
    
    # Create and use product A
    product_a = factory.create_product("A")
    print(product_a.operation())
    
    # Create and use product B
    product_b = factory.create_product("B")
    print(product_b.operation())

# ============= FACTORY METHOD EXAMPLE =============

class Creator(ABC):
    """Creator class with factory method"""
    @abstractmethod
    def factory_method(self):
        """Factory method to be implemented by subclasses"""
        pass
    
    def some_operation(self):
        """Business logic that uses the factory method"""
        # Call the factory method to create a Product object
        product = self.factory_method()
        
        # Use the product
        result = f"Creator: The same creator's code has just worked with {product.operation()}"
        return result

class ConcreteCreator1(Creator):
    """Concrete creator that returns ConcreteProductA"""
    def factory_method(self):
        return ConcreteProductA()

class ConcreteCreator2(Creator):
    """Concrete creator that returns ConcreteProductB"""
    def factory_method(self):
        return ConcreteProductB()

# Client code for Factory Method
def factory_method_client():
    """Demo of Factory Method Pattern"""
    print("\n=== Factory Method Example ===")
    
    print("App: Launched with ConcreteCreator1")
    client_code(ConcreteCreator1())
    
    print("\nApp: Launched with ConcreteCreator2")
    client_code(ConcreteCreator2())

def client_code(creator):
    """Client code that works with any creator subclass"""
    print(f"Client: I'm not aware of the creator's class, but it still works.\n"
          f"{creator.some_operation()}")

# ============= ABSTRACT FACTORY EXAMPLE =============

class AbstractProductA(ABC):
    """Abstract product A interface"""
    @abstractmethod
    def useful_function_a(self):
        pass

class ConcreteProductA1(AbstractProductA):
    """Concrete product A1 - Family 1"""
    def useful_function_a(self):
        return "The result of the product A1."

class ConcreteProductA2(AbstractProductA):
    """Concrete product A2 - Family 2"""
    def useful_function_a(self):
        return "The result of the product A2."

class AbstractProductB(ABC):
    """Abstract product B interface"""
    @abstractmethod
    def useful_function_b(self):
        pass
    
    @abstractmethod
    def another_useful_function_b(self, collaborator: AbstractProductA):
        """
        B products can work with A products
        """
        pass

class ConcreteProductB1(AbstractProductB):
    """Concrete product B1 - Family 1"""
    def useful_function_b(self):
        return "The result of the product B1."
    
    def another_useful_function_b(self, collaborator: AbstractProductA):
        result = collaborator.useful_function_a()
        return f"The result of B1 collaborating with ({result})"

class ConcreteProductB2(AbstractProductB):
    """Concrete product B2 - Family 2"""
    def useful_function_b(self):
        return "The result of the product B2."
    
    def another_useful_function_b(self, collaborator: AbstractProductA):
        result = collaborator.useful_function_a()
        return f"The result of B2 collaborating with ({result})"

class AbstractFactory(ABC):
    """Abstract Factory Interface"""
    @abstractmethod
    def create_product_a(self) -> AbstractProductA:
        pass
    
    @abstractmethod
    def create_product_b(self) -> AbstractProductB:
        pass

class ConcreteFactory1(AbstractFactory):
    """Concrete Factory for Family 1 products"""
    def create_product_a(self) -> AbstractProductA:
        return ConcreteProductA1()
    
    def create_product_b(self) -> AbstractProductB:
        return ConcreteProductB1()

class ConcreteFactory2(AbstractFactory):
    """Concrete Factory for Family 2 products"""
    def create_product_a(self) -> AbstractProductA:
        return ConcreteProductA2()
    
    def create_product_b(self) -> AbstractProductB:
        return ConcreteProductB2()

def abstract_factory_client():
    """Demo of Abstract Factory Pattern"""
    print("\n=== Abstract Factory Example ===")
    
    print("Client: Testing client code with the first factory type:")
    client_abstract_factory_code(ConcreteFactory1())
    
    print("\nClient: Testing the same client code with the second factory type:")
    client_abstract_factory_code(ConcreteFactory2())

def client_abstract_factory_code(factory: AbstractFactory):
    """
    Client code that works with factories and products through abstract interfaces
    """
    product_a = factory.create_product_a()
    product_b = factory.create_product_b()
    
    print(f"{product_b.useful_function_b()}")
    print(f"{product_b.another_useful_function_b(product_a)}")


# ============= REAL-WORLD EXAMPLE: UI COMPONENTS =============

# Abstract products
class Button(ABC):
    @abstractmethod
    def render(self):
        pass
    
    @abstractmethod
    def on_click(self):
        pass

class Checkbox(ABC):
    @abstractmethod
    def render(self):
        pass
    
    @abstractmethod
    def toggle(self):
        pass

# Concrete products for Windows
class WindowsButton(Button):
    def render(self):
        return "Rendering a Windows button"
    
    def on_click(self):
        return "Windows button clicked!"

class WindowsCheckbox(Checkbox):
    def render(self):
        return "Rendering a Windows checkbox"
    
    def toggle(self):
        return "Windows checkbox toggled!"

# Concrete products for Web
class WebButton(Button):
    def render(self):
        return "Rendering a button in HTML"
    
    def on_click(self):
        return "JavaScript click event triggered!"

class WebCheckbox(Checkbox):
    def render(self):
        return "Rendering a checkbox in HTML"
    
    def toggle(self):
        return "JavaScript toggle event triggered!"

# Abstract factory
class GUIFactory(ABC):
    @abstractmethod
    def create_button(self) -> Button:
        pass
    
    @abstractmethod
    def create_checkbox(self) -> Checkbox:
        pass

# Concrete factories
class WindowsFactory(GUIFactory):
    def create_button(self) -> Button:
        return WindowsButton()
    
    def create_checkbox(self) -> Checkbox:
        return WindowsCheckbox()

class WebFactory(GUIFactory):
    def create_button(self) -> Button:
        return WebButton()
    
    def create_checkbox(self) -> Checkbox:
        return WebCheckbox()

# Client code
class Application:
    def __init__(self, factory: GUIFactory):
        self._factory = factory
        self._button = None
        self._checkbox = None
    
    def create_ui(self):
        self._button = self._factory.create_button()
        self._checkbox = self._factory.create_checkbox()
        
        print("Creating UI:")
        print(self._button.render())
        print(self._checkbox.render())
    
    def click_button(self):
        print(self._button.on_click())
    
    def toggle_checkbox(self):
        print(self._checkbox.toggle())

def real_world_example():
    """Demo of a real-world Abstract Factory example"""
    print("\n=== Real-World UI Factory Example ===")
    
    # Determine which factory to use based on configuration or environment
    if sys.platform.startswith('win'):
        print("\nRunning on Windows, using Windows UI components")
        factory = WindowsFactory()
    else:
        print("\nRunning on a non-Windows platform, using Web UI components")
        factory = WebFactory()
    
    # Create and use the application
    app = Application(factory)
    app.create_ui()
    app.click_button()
    app.toggle_checkbox()


if __name__ == "__main__":
    print("Factory Pattern Demonstrations")
    print("==============================")
    
    # Run the Simple Factory example
    simple_factory_client()
    
    # Run the Factory Method example
    factory_method_client()
    
    # Run the Abstract Factory example
    abstract_factory_client()
    
    # Run the real-world example
    real_world_example() 